platform/android/java/lib/src/org/godotengine/godot/tts/GodotTTS.java

```diff
@@ -0,0 +1,299 @@
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+package org.godotengine.godot.tts;
+
+import org.godotengine.godot.GodotLib;
+
+import android.app.Activity;
+import android.os.Bundle;
+import android.speech.tts.TextToSpeech;
+import android.speech.tts.UtteranceProgressListener;
+import android.speech.tts.Voice;
+
+import androidx.annotation.Keep;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Set;
+
+/**
+ * Wrapper for Android Text to Speech API and custom utterance query implementation.
+ * <p>
+ * A [GodotTTS] provides the following features:
+ * <p>
+ * <ul>
+ * <li>Access to the Android Text to Speech API.
+ * <li>Utterance pause / resume functions, unsupported by Android TTS API.
+ * </ul>
+ */
+@Keep
+public class GodotTTS extends UtteranceProgressListener {
+	// Note: These constants must be in sync with DisplayServer::TTSUtteranceEvent enum from "servers/display_server.h".
+	final private static int EVENT_START = 0;
+	final private static int EVENT_END = 1;
+	final private static int EVENT_CANCEL = 2;
+	final private static int EVENT_BOUNDARY = 3;
+
+	final private TextToSpeech synth;
+	final private LinkedList<GodotUtterance> queue;
+	final private Object lock = new Object();
+	private GodotUtterance lastUtterance;
+
+	private boolean speaking;
+	private boolean paused;
+
+	public GodotTTS(Activity p_activity) {
+		synth = new TextToSpeech(p_activity, null);
+		queue = new LinkedList<GodotUtterance>();
+
+		synth.setOnUtteranceProgressListener(this);
+	}
+
+	private void updateTTS() {
+		if (!speaking && queue.size() > 0) {
+			int mode = TextToSpeech.QUEUE_FLUSH;
+			GodotUtterance message = queue.getFirst();
+
+			Set<Voice> voices = synth.getVoices();
+			for (Voice v : voices) {
+				if (v.getName() == message.voice) {
+					synth.setVoice(v);
+					break;
+				}
+			}
+			synth.setPitch(message.pitch);
+			synth.setSpeechRate(message.rate);
+
+			Bundle params = new Bundle();
+			params.putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, message.volume / 100.f);
+
+			lastUtterance = message;
+			lastUtterance.start = 0;
+			lastUtterance.offset = 0;
+			paused = false;
+
+			synth.speak(message.text, mode, params, String.valueOf(message.id));
+			speaking = true;
+			queue.pollFirst();
+		}
+	}
+
+	/**
+	 * Called by TTS engine when the TTS service is about to speak the specified range.
+	 */
+	@Override
+	public void onRangeStart(String utteranceId, int start, int end, int frame) {
+		synchronized (lock) {
```

Comment:
Here and below, is the use of the `lock` because you assume these methods can be invoked from several threads (e.g: main vs render thread)?

If that's the case, we need to take a different approach as the usage of `lock` is by itself insufficient.

Reply:
I have no idea how Android platform is implemented and what's called from what thread (it seems to be called from the rendering thread only). So added it just as a precaution by analogy with `_THREAD_SAFE_METHOD_` on other platforms.

Reply:
I can push the necessary changes to your branch later. Do you have a sample project I can use to validate the changes?

For reference on Android platforms, the Godot Engine runtime is spread between the main and render thread. The render thread is where the core of the engine runs, while the main thread is used for initial setup (`Godot.java`), input handling, and any other components/logic that interface with the underlying Android platform APIs.

In this PR, the `tts` calls are originating from the engine, and thus are running on the render thread. 
They cross the jni boundary and interact with the Android `TextToSpeech` api, so one of the first thing to check is to make sure that api can be called from the render thread.

In return the `TextToSpeech` api provides callbacks to notify of the state of the `tts` request, and these callbacks are routed back to the engine. So here we need to check which thread the `TextToSpeech` api uses to dispatch these callbacks and ensure they're routed back to the engine on the proper (render) thread.

Let me know if you have any questions.

Reply:
> Do you have a sample project I can use to validate the changes?

There's a test project in the first post - https://github.com/godotengine/godot/files/7788994/tts2_test.zip

>  Android TextToSpeech api, so one of the first thing to check is to make sure that api can be called from the render thread.

I do not see anything in the [documentation](https://developer.android.com/reference/android/speech/tts/TextToSpeech#synthesizeToFile(java.lang.CharSequence,%20android.os.Bundle,%20android.os.ParcelFileDescriptor,%20java.lang.String)), and exported project works fine on my device, so I assume it can be called from any thread.

> In return the TextToSpeech api provides callbacks to notify of the state of the tts request, and these callbacks are routed back to the engine. So here we need to check which thread the TextToSpeech api uses to dispatch these callbacks and ensure they're routed back to the engine on the proper (render) thread.

[The callbacks specified in this method can be called from multiple threads](https://developer.android.com/reference/android/speech/tts/UtteranceProgressListener).

> I can push the necessary changes to your branch later.

Sure.